### PR TITLE
fix #5794 fix #5755 fix #5759 feat(nimbus): add Clone Experiment dialog summoned via sidebar action button

### LINKED ISSUE #5755: (FE) Create modal to clone experiment
Using the new modal component introduced in EXP-1370 create a new instance of it that contains:

* The title “Clone [experiment name]”
* A single text field to input the experiment name, prefilled with the text “[experiment name] Copy”
* The new slug preview component created in EXP-1371
* Buttons to Cancel and Save
* A space at the top of the modal body to display server errors

This new component should also have a new GQL query set up that calls the clone mutation created in EXP-1369.

Clicking Cancel should abort the operation and not clone the experiment. Clicking Save should perform the mutation, wait for a successful response, and then redirect to that newly-created experiment. Server errors should be displayed if they occur.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1373)

### LINKED ISSUE #5759: (FE) Add sidebar button to trigger experiment clone modal
Referring to this [Figma doc](https://www.figma.com/file/5SLa07XqQb0ttB1ay1gami/Nimbus-Console-UI?node-id=219194%3A107743), create a new section at the bottom of [AppLayoutWithSidebar](https://github.com/mozilla/experimenter/blob/main/app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.tsx#L58:L58)*, containing a button for cloning the experiment. Attached to this issue is the icon you should use.

The button should always be visible and when clicked should open the clone modal created in EXP-1373. The label should say “Clone this experiment”.

* EXP-1296 also adds a new button to this area of the sidebar for archiving, so depending on when this issue is worked on this new section may already be present.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1374)
┆Attachments: <a href="https://mozilla-hub.atlassian.net/secure/attachment/14663/clone.svg">clone.svg</a>

### LINKED ISSUE #5794: (FE) Create experiment slug preview component
Create a new component that takes an experiment “name” string and displays a parameterized / slugified version of it. This is meant to simply display a preview of what an experiment’s slug will be. Include a :question: icon at the end of it, with the tooltip text “The slug cannot be changed once an experiment has been created”.

Update the [FormOverview](https://github.com/mozilla/experimenter/blob/main/app/experimenter/nimbus-ui/src/components/FormOverview/index.tsx#L57:L57) component to render this component when creating a new experiment. It will also be used in the cloning modal.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1371)

### PR DESCRIPTION
**Pull request by lmorchard**
Because:

- We want to remove friction by allowing users to clone experiments

This commit:

- adds a new CloneDialog sub-component for SidebarActions that accepts a
  new name based on the current experiment and offers a slug preview

- adds a Clone Experiment button in SidebarActions to summon CloneDialog

- adds SlugTextControl, a small component that shows a slugified preview
  of string value

### CODE REVIEW

**Comment by lmorchard** [comment]
Opening as a draft until I see CI go green. Could be broken up, but seemed like that would be two tiny PRs and one still pretty-big PR.

**Comment by jaredlockhart** [comment]
fix fix fix!

**Comment by lmorchard** [comment]
Some story links:

* [Standalone CloneDialog, visible](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/c090fb0a0f77e2070f8d4e98784867f3c95f0257/nimbus-ui/index.html?path=/story/components-sidebaractions-clonedialog--shown)
* [SidebarActions with the new Clone Experiment button](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/c090fb0a0f77e2070f8d4e98784867f3c95f0257/nimbus-ui/index.html?path=/story/components-sidebaractions--archive-enabled)
* [Draft experiment with the Clone Experiment button available](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/c090fb0a0f77e2070f8d4e98784867f3c95f0257/nimbus-ui/index.html?path=/story/pages-summary--draft-status)

Not sure if we'll want to merge this before the backend. But, just in case, I threw a "not implemented yet" alert in there

**Comment by lmorchard** [comment]
> fix fix fix!

My favorite song off Paramore's 2007 album "Riot!"

**Comment by jaredlockhart** [comment]
> Some story links:
> 
>     * [Standalone CloneDialog, visible](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/c090fb0a0f77e2070f8d4e98784867f3c95f0257/nimbus-ui/index.html?path=/story/components-sidebaractions-clonedialog--shown)
> 
>     * [SidebarActions with the new Clone Experiment button](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/c090fb0a0f77e2070f8d4e98784867f3c95f0257/nimbus-ui/index.html?path=/story/components-sidebaractions--archive-enabled)
> 
>     * [Draft experiment with the Clone Experiment button available](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/c090fb0a0f77e2070f8d4e98784867f3c95f0257/nimbus-ui/index.html?path=/story/pages-summary--draft-status)
> 
> 
> Not sure if we'll want to merge this before the backend. But, just in case, I threw a "not implemented yet" alert in there

OOooooo looking good!  You can take the word `Experiment` off the clone button, I realized it was redundant and took it off the archive button, don't get that ugly wrap around that way.

**Comment by lmorchard** [comment]
Looks like some significant PR landing, so I rebased against main and pushed again. Don't *think* I broke anything with merge conflict resolutions

**Comment by lmorchard** [comment]
Gonna open for review, fire away with any further needful changes

File: app/experimenter/nimbus-ui/src/components/SidebarActions/CloneDialog/index.tsx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+  const handleSave = handleSubmit(onSave);
+
+  const nameControlAttrs = formControlAttrs("name", {
+    ...REQUIRED_FIELD,
+    maxLength: {
+      value: 80,
+      message: "Cannot be greater than 80 characters",
+    },
+  });
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/SidebarActions/CloneDialog/index.tsx:68]: We have another ticket https://github.com/mozilla/experimenter/issues/5600 open to go in and validate all the different identifiers that make their way out in the DTO, so we'll have to go in and tighten all the validation on these in the serializers, so I'd leave this out of here and just rely on the server validation.

File: app/experimenter/nimbus-ui/src/components/SidebarActions/CloneDialog/index.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+          </Form>
+        </FormProvider>
+      </Modal.Body>
+      <Modal.Footer>
+        <Button variant="secondary" disabled={isLoading} onClick={onCancel}>
+          Cancel
+        </Button>
+        <Button
+          variant="primary"
+          disabled={isLoading || !isValid}
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/SidebarActions/CloneDialog/index.tsx:124]: Don't block the button on valid 🙏

File: app/experimenter/nimbus-ui/src/components/SlugTextControl/index.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+export const SlugTextControl = ({
+  value,
+  defaultValue,
+}: SlugTextControlProps) => (
+  <Form.Control
+    readOnly
+    data-testid="SlugTextControl"
+    value={slugify(value || defaultValue)}
+    type="text"
+  />
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/SlugTextControl/index.tsx:23]: Oh yeah interesting I've also heard people ask about this for the regular create form.  This is great after we land this we should also backport this to the regular form.

File: app/experimenter/nimbus-ui/src/components/SlugTextControl/slugify.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  if (!text) return "";
+  return text
+    .toString()
+    .toLowerCase()
+    .replace(/\s+/g, "-") // Replace spaces with -
+    .replace(/[^\w-]+/g, "") // Remove all non-word chars
+    .replace(/--+/g, "-") // Replace multiple - with single -
+    .replace(/^[-_]+/, "") // Trim - from start of text
+    .replace(/[-_]+$/, ""); // Trim - from end of text
+}
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/SlugTextControl/slugify.ts:15]: Oh interesting........ I think the serializer currently expects that we don't explicitly pass in the slug and it's auto generated....

Yep: https://github.com/mozilla/experimenter/blob/main/app/experimenter/experiments/api/v5/serializers.py#L576

So... when we go and hook this up we can either let the frontend to the slug and bypass that on the server, or only pass name and let the server do the slugification and just let this be a handy preview but then in theory they might not match...  Yeah I dunno couple possibilities.

**Review by jaredlockhart — CHANGES_REQUESTED** [review]
This is a damn fine modal.  I don't think we should have the validation on the client, but other than that everything looks good.  The slug is something to think about but probably either way we do it will be fine.

File: app/experimenter/nimbus-ui/src/components/SlugTextControl/slugify.ts (hunk 4 above)
Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/components/SlugTextControl/slugify.ts:15]: Oh yeah, this doesn't get submitted at all. This was just in the design as a preview. Tried to keep it close to Django's impl, but either way the server wins. This value doesn't even get passed back in the `onSave` callback to SidebarActions

File: app/experimenter/nimbus-ui/src/components/SlugTextControl/index.tsx (hunk 3 above)
Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/components/SlugTextControl/index.tsx:23]: Yeah, should be close to a one-liner plus tests. Just a small component that takes a value and shows a slugified preview

**Comment by lmorchard** [comment]
Stripped back the validation requirements, which simplifies some bits

**Review by jaredlockhart — APPROVED** [review]
Looks good 👍

**Comment by lmorchard** [comment]
Alright, I'm gonna land this

**Pull request merged**